compiler/rustc_expand/src/proc_macro_server.rs

```diff
@@ -421,13 +425,58 @@ impl server::TokenStream for Rustc<'_> {
         parse_stream_from_source_str(
             FileName::proc_macro_source_code(src),
             src.to_string(),
-            self.sess,
+            self.sess(),
             Some(self.call_site),
         )
     }
     fn to_string(&mut self, stream: &Self::TokenStream) -> String {
         pprust::tts_to_string(stream)
     }
+    fn expand_literal(&mut self, stream: &Self::TokenStream) -> Result<Self::Literal, ()> {
+        // Parse the expression from our tokenstream.
+        let expr = {
+            let mut p = rustc_parse::stream_to_parser(
+                self.sess(),
+                stream.clone(),
+                Some("proc_macro expand literal"),
+            );
+            let expr = p.parse_expr().map_err(|mut err| {
+                err.cancel();
```

Comment:
After some thought, perhaps it's better to report the error.
It's a strictly more conservative choice (it produces a compilation error), it makes it visible why `expand_literal` failed, and we can relax it later if convincing use cases are provided.

Reply:
Should we report an error on every error return path so that callers can't rely on any failure not reporting an error & failing compilation to be even more conservative, or just report the parse error?

With just reporting an error on expansion failure or parse errors, I could see callers relying on e.g. checking if an expression is literal by calling `expand_literal`, and assuming it isn't literal & using it as a runtime expression with a different expansion if `expand_literal` fails. This could theoretically be used to support something like a `regex!` with a potentially dynamic string argument which eagerly compiles if the expression is a literal.

I don't really see much use for recovering on parse errors other than potentially avoiding duplicate errors if the macro both checks if it's a literal and uses it as a normal expression (though a macro should be able to pre-parse the expression with `syn`)

Reply:
>Should we report an error on every error return path so that callers can't rely on any failure not reporting an error & failing compilation to be even more conservative, or just report the parse error?

I'm going to leave this for the libs team to decide (since this PR is `T-libs-api` now).
I'd be ok with always reporting an error for now.